internal/controller/prefectworkpool_controller.go

```diff
@@ -52,81 +53,63 @@ type PrefectWorkPoolReconciler struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.3/pkg/reconcile
 func (r *PrefectWorkPoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	var condition metav1.Condition
-	var err error
-
 	log := ctrllog.FromContext(ctx)
 
 	log.Info("Reconciling PrefectWorkPool")
 
 	workPool := &prefectiov1.PrefectWorkPool{}
-	err = r.Get(ctx, req.NamespacedName, workPool)
+	err := r.Get(ctx, req.NamespacedName, workPool)
 	if errors.IsNotFound(err) {
 		return ctrl.Result{}, nil
 	} else if err != nil {
 		return ctrl.Result{}, err
 	}
 
+	// Defer a final status update at the end of the reconciliation loop, so that any of the
+	// individual reconciliation functions can update the status as they see fit.
 	defer func() {
-		if updateErr := r.updateCondition(ctx, workPool, condition); updateErr != nil {
-			err = utilerrors.NewAggregate([]error{updateErr, err})
+		if statusErr := r.Status().Update(ctx, workPool); statusErr != nil {
+			log.Error(statusErr, "Failed to update WorkPool status")
 		}
 	}()
 
+	objName := constants.Deployment
+
 	desiredDeployment := r.prefectWorkerDeployment(workPool)
 
-	workPoolNamespacedName := types.NamespacedName{
-		Namespace: workPool.Namespace,
-		Name:      workPool.Name,
+	deploy := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      desiredDeployment.Name,
+			Namespace: desiredDeployment.Namespace,
+		},
 	}
 
-	foundDeployment := &appsv1.Deployment{}
-	err = r.Get(ctx, workPoolNamespacedName, foundDeployment)
-	if errors.IsNotFound(err) {
-		log.Info("Creating Deployment", "name", desiredDeployment.Name)
-		if err = r.Create(ctx, &desiredDeployment); err != nil {
-			condition = conditions.NotCreated("Deployment", err)
-			return ctrl.Result{}, err
+	result, err := controllerutil.CreateOrUpdate(ctx, r.Client, deploy, func() error {
+		if err := ctrl.SetControllerReference(workPool, deploy, r.Scheme); err != nil {
+			return err
 		}
-		condition = conditions.Created("Deployment")
-	} else if err != nil {
-		condition = conditions.UnknownError("Deployment", err)
+
+		return mergo.Merge(deploy, desiredDeployment, mergo.WithOverride)
+	})
+
+	log.Info("CreateOrUpdate", "object", objName, "name", desiredDeployment.Name, "result", result)
+
+	if err != nil {
+		meta.SetStatusCondition(&workPool.Status.Conditions, conditions.UnknownError(objName, err))
 		return ctrl.Result{}, err
-	} else if !metav1.IsControlledBy(foundDeployment, workPool) {
```

Comment:
are we skipping this check now bc it's handled already in the CreateOrUpdate utility?

Reply:
Yeah exactly - I left some breadcrumbs in https://github.com/PrefectHQ/prefect-operator/pull/55#issuecomment-2319007634.

Reply:
makes sense